### DISCUSSION TITLE
How to write a sensor to gather asset execution metrics like start time?

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#17536] [Q&A] [answered]
We've been using `run_status_sensor` sensors to monitor the start, success and failure of jobs in our dagster deployment and then sending metrics to DataDog via statsd. We'd like to do something similar with asset level metrics. I read about `multi_asset_sensor` which might let us select assets to monitor, but I haven't been able to find metrics like asset start time on the `MultiAssetSensorEvaluationContext`. Am I looking in the wrong place or is there a better way to gather metrics like runtime for assets?

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1698775210504939?thread_ts=1698775210.504939&cid=C01U954MEER).*

**Comment by sryza** [comment, marked as answer]
Here's an example of a multi-asset sensor that collects the run times of the steps that materialized a particular set of assets, since the last sensor tick:

```python
from dagster import multi_asset_sensor


@multi_asset_sensor(monitored_assets=...)
def gather_asset_metrics(context):
    materialization_records = [
        record
        for asset_key in context.asset_keys
        for record in context.materialization_records_for_key(asset_key)
    ]
    run_ids = {record.run_id for record in materialization_records}
    materialization_run_ids = {
        (record.run_id, record.asset_key) for record in materialization_records
    }

    for run_id in run_ids:
        run_step_stats = context.instance.get_run_step_stats(run_id)
        for step_stats in run_step_stats:
            for event in step_stats.materialization_events:
                if (event.run_id, event.asset_key) in materialization_run_ids:
                    runtime = step_stats.end_time - step_stats.start_time
                    ...

    context.advance_all_cursors()
```

**Reply by nathanskone** [reply]
@sryza Is this the recommended path to get job/asset metrics out of Dagster into external monitoring tools? Are there other methods or approaches that are better?

**Reply by sryza** [reply]
@nathanskone - for pushing metrics into external monitoring tools, using a sensor is the recommended path.

**Reply by rover-wagnar** [reply]
Thank you @sryza, this is great!

**Comment by rover-wagnar** [comment]
Would you help me take another look at this, @sryza?
When I ran this code in tests of materializing assets, I was getting continuous events and metrics instead of discrete events and metrics. It looks like what's happening is the cursor updates to the latest materialization events but not passed them, so we end up sending metrics for the same materialization events whenever the `gather_asset_metrics` function runs.

I was expecting `context.advance_all_cursors()` to update the cursor so that successive runs of the `gather_asset_metrics` method would not see the same asset materializations, but if I log the method runs I'm seeing previous events being reprocessed.

I tried both changing to using `context.latest_materialization_records_by_key()` and also manually setting a cursor that has a serialized state for tracking previously processed AssetKey_RunId combinations, but no luck. It seems like (1) user-supplied cursors are getting overwritten between runs of the `gather_asset_metrics` method and (2) `context.advance_all_cursors()` isn't marking materialization events as "consumed" and they're being fetched in a loop in `gather_asset_metrics` until a new asset materialization is triggered.

Am I misunderstanding how the cursor works? I want to process an AssetKey_RunId combination only once for metrics and events and it seems like I need some kind of state management to make that happen.

**Reply by sryza** [reply]
I don't have time to look into this deeply right now, but if you're having trouble with the cursor for multi-asset sensors, my recommendation would be to use a regular sensor instead, which doesn't do complex cursor management behind the scenes.  This will give you more control and make the behavior more transparent.

Instead of invoking `context.materialization_records_for_key`, you would invoke the underlying instance methods: https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/definitions/multi_asset_sensor_definition.py#L418.

**Reply by rover-wagnar** [reply]
Awesome, thank you for the suggestion @sryza ! I was indeed able to properly set the cursor and track state between sensor evaluations using a normal `@sensor`.

Follow-up questions for using the following API to fetch Materialization Events:
```
# context
# consumed_asset_events is a mapping[AssetKey, (latest) event storage_id] to track the latest processed events for each Asset being tracked

materialization_records = context.instance.get_event_records(
        EventRecordsFilter(
            event_type=DagsterEventType.ASSET_MATERIALIZATION,
            # Don't fetch events before the smallest storage_id being tracked for an Asset's materialization events,
            #   because these are earlier materialization events that should have already been consumed.
            after_cursor=min(consumed_asset_events.values()) if consumed_asset_events.values() else 0
        )
    )
```

(1) Can I assume materialization event `storage_id` values are always increasing? and (2) how distinct are `storage_id` values? [The code in `MultiAssetSensorContext`](https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/definitions/multi_asset_sensor_definition.py#L441) suggets using the latest_consumed_event_id for `after_cursor`, which suggests increasing and distinct values for `storage_id`, but I'm not sure if `storage_id` is unique across assets or partitions.
